File: cpp/src/SculptEngine.cpp
#include "SculptEngine.hpp"
#include <algorithm>
#include <cmath>

namespace {
float moveFalloff(float distance, float radius) {
  const float t = std::clamp(1.0f - distance / radius, 0.0f, 1.0f);
  return t * t * (3.0f - 2.0f * t);
}
}

SculptEngine::SculptEngine() { createQuadSphere(1.0f, 5); }
void SculptEngine::createQuadSphere(float radius, int level) { mesh_ = PrimitiveGenerator::quadSphere(radius, level); coarseLevels_.clear(); moveGrabbed_.clear(); }
void SculptEngine::subdivideCurrent() { coarseLevels_.push_back(mesh_); PrimitiveGenerator::subdivideCurrent(mesh_); }
bool SculptEngine::restoreCoarseLevel() {
  if (coarseLevels_.empty()) return false;
  mesh_ = std::move(coarseLevels_.back());
  coarseLevels_.pop_back();
  return true;
}
bool SculptEngine::applyDraw(float x,float y,float z,float radius,float strength,bool invert) {
  return Brush::draw(mesh_, {x,y,z}, {radius,strength,invert});
}
bool SculptEngine::applySmooth(float x,float y,float z,float radius,float strength) {
  return Brush::smooth(mesh_, {x,y,z}, {radius,strength,false});
}
bool SculptEngine::applyClay(float x,float y,float z,float nx,float ny,float nz,float radius,float strength,bool invert) {
  return Brush::clay(mesh_, {x,y,z}, {nx,ny,nz}, {radius,strength,invert});
}
void SculptEngine::beginMove(float x,float y,float z,float radius) {
  const Vec3 center{x,y,z};
  const float radius_squared = radius * radius;
  moveGrabbed_.clear();
  for (std::size_t i = 0; i < mesh_.vertices.size(); ++i) {
    const Vec3 offset = mesh_.vertices[i] - center;
    const float distance_squared = offset.lengthSquared();
    if (distance_squared > radius_squared) continue;
    moveGrabbed_.push_back({
      static_cast<std::uint32_t>(i),
      moveFalloff(std::sqrt(distance_squared), radius),
      mesh_.vertices[i],
    });
  }
}
bool SculptEngine::applyMove(float dx,float dy,float dz,bool invert) {
  if (moveGrabbed_.empty()) return false;
  const float direction = invert ? -1.0f : 1.0f;
  const Vec3 delta{dx * direction, dy * direction, dz * direction};
  for (const auto& vertex : moveGrabbed_) {
    mesh_.vertices[vertex.index] = vertex.base + delta * vertex.weight;
  }
  mesh_.rebuildDerivedData();
  return true;
}
bool SculptEngine::pixRemesh(const std::vector<float>& positions,
                             const std::vector<std::uint32_t>& indices,
                             int resolution,
                             float adaptiveDensity,
                             bool preserveSharpFeatures,
                             int smoothIterations,
                             bool projectDetails) {
  if (positions.size() < 9 || indices.size() < 3) return false;
  Mesh source;
  source.vertices.reserve(positions.size() / 3);
  for (std::size_t index = 0; index + 2 < positions.size(); index += 3) {
    source.vertices.push_back({positions[index], positions[index + 1], positions[index + 2]});
  }
  for (std::size_t index = 0; index + 2 < indices.size(); index += 3) {
    if (indices[index] >= source.vertices.size() ||
        indices[index + 1] >= source.vertices.size() ||
        indices[index + 2] >= source.vertices.size()) {
      continue;
    }
    source.triangles.push_back({indices[index], indices[index + 1], indices[index + 2]});
  }
  source.rebuildDerivedData();

  PixRemeshOptions options;
  options.resolution = resolution;
  options.adaptiveDensity = adaptiveDensity;
  options.preserveSharpFeatures = preserveSharpFeatures;
  options.smoothIterations = smoothIterations;
  options.projectDetails = projectDetails;

  PixRemesh remesher;
  Mesh result = remesher.remesh(source, options);
  if (result.vertices.empty() || result.indices.size() < 12 || result.vertices.size() > result.indices.size()) return false;
  history_.push(mesh_);
  mesh_ = std::move(result);
  coarseLevels_.clear();
  moveGrabbed_.clear();
  return true;
}
std::uint32_t SculptEngine::previewPixRemeshTriangles(const std::vector<float>& positions,
                                                      const std::vector<std::uint32_t>& indices,
                                                      int resolution,
                                                      float adaptiveDensity,
                                                      bool preserveSharpFeatures,
                                                      int smoothIterations,
                                                      bool projectDetails) {
  if (positions.size() < 9 || indices.size() < 3) return 0;
  Mesh source;
  source.vertices.reserve(positions.size() / 3);
  for (std::size_t index = 0; index + 2 < positions.size(); index += 3) {
    source.vertices.push_back({positions[index], positions[index + 1], positions[index + 2]});
  }
  for (std::size_t index = 0; index + 2 < indices.size(); index += 3) {
    if (indices[index] >= source.vertices.size() ||
        indices[index + 1] >= source.vertices.size() ||
        indices[index + 2] >= source.vertices.size()) {
      continue;
    }
    source.triangles.push_back({indices[index], indices[index + 1], indices[index + 2]});
  }
  source.rebuildDerivedData();
  PixRemeshOptions options;
  options.resolution = resolution;
  options.adaptiveDensity = adaptiveDensity;
  options.preserveSharpFeatures = preserveSharpFeatures;
  options.smoothIterations = smoothIterations;
  options.projectDetails = projectDetails;
  const auto stats = PixRemesh{}.preview(source, options);
  return stats.vertices > stats.triangles * 3 || stats.triangles < 4 ? 0 : stats.triangles;
}
void SculptEngine::beginStroke() { history_.push(mesh_); }
bool SculptEngine::undo() { return history_.undo(mesh_); }
bool SculptEngine::redo() { return history_.redo(mesh_); }

File: src/wasm/SculptEngine.ts
import * as THREE from 'three';
import type { BrushSettings } from '../sculpt/Brush';
import { DrawBrush } from '../sculpt/brushes/DrawBrush';
import { SmoothBrush } from '../sculpt/brushes/SmoothBrush';
import { ClayBrush } from '../sculpt/brushes/ClayBrush';
import { MoveBrush } from '../sculpt/brushes/MoveBrush';
import { buildVertexNeighbors, queryBrushCandidates } from '../utils/GeometryUtils';

interface EmbindVector<T> {
  size(): number;
  get(index: number): T;
  delete(): void;
}

interface MutableEmbindVector<T> extends EmbindVector<T> {
  push_back(value: T): void;
}

interface WasmEngine {
  createQuadSphere(radius: number, subdivisionLevel: number): void;
  subdivideCurrent(): void;
  restoreCoarseLevel(): boolean;
  pixRemesh(positions: MutableEmbindVector<number>, indices: MutableEmbindVector<number>, resolution: number, adaptiveDensity: number, preserveSharpFeatures: boolean, smoothIterations: number, projectDetails: boolean): boolean;
  previewPixRemeshTriangles(positions: MutableEmbindVector<number>, indices: MutableEmbindVector<number>, resolution: number, adaptiveDensity: number, preserveSharpFeatures: boolean, smoothIterations: number, projectDetails: boolean): number;
  applyDraw(x: number, y: number, z: number, radius: number, strength: number, invert: boolean): boolean;
  applySmooth(x: number, y: number, z: number, radius: number, strength: number): boolean;
  applyClay(x: number, y: number, z: number, nx: number, ny: number, nz: number, radius: number, strength: number, invert: boolean): boolean;
  beginMove(x: number, y: number, z: number, radius: number): void;
  applyMove(dx: number, dy: number, dz: number, invert: boolean): boolean;
  beginStroke(): void;
  positions(): EmbindVector<number>;
  normals(): EmbindVector<number>;
  indices(): EmbindVector<number>;
  displayEdges(): EmbindVector<number>;
  quads(): EmbindVector<number>;
}

interface WasmModule {
  SculptEngine: new () => WasmEngine;
  FloatVector: new () => MutableEmbindVector<number>;
  UIntVector: new () => MutableEmbindVector<number>;
}

export class SculptEngine {
  onReady: (() => void) | undefined;
  private wasm: WasmEngine | undefined;
  private wasmModule: WasmModule | undefined;
  private loadError: string | undefined;
  private readonly drawFallback = new DrawBrush();
  private readonly smoothFallback = new SmoothBrush();
  private readonly clayFallback = new ClayBrush();
  private readonly moveFallback = new MoveBrush();

  constructor(private readonly getGeometry: () => THREE.BufferGeometry) {
    void this.load();
  }

  createQuadSphere(radius: number, subdivisionLevel: number): void {
    if (!this.wasm) return;
    this.wasm.createQuadSphere(radius, subdivisionLevel);
  }

  beginStroke(): void {
    this.wasm?.beginStroke();
  }

  isWasmActive(): boolean {
    return Boolean(this.wasm);
  }

  getWasmLoadError(): string | undefined {
    return this.loadError;
  }

  subdivideCurrent(): void {
    if (!this.wasm) return;
    this.wasm.subdivideCurrent();
  }

  restoreCoarseLevel(): void {
    if (!this.wasm || !this.wasm.restoreCoarseLevel()) return;
  }

  pixRemesh(
    positions: number[],
    indices: number[],
    options: { resolution: number; adaptiveDensity: number; preserveSharpFeatures: boolean; smoothIterations: number; projectDetails: boolean },
  ): THREE.BufferGeometry | undefined {
    if (!this.wasm || !this.wasmModule) return undefined;
    const buffers = this.createWasmBuffers(positions, indices);
    try {
      const changed = this.wasm.pixRemesh(
        buffers.positions,
        buffers.indices,
        options.resolution,
        options.adaptiveDensity,
        options.preserveSharpFeatures,
        options.smoothIterations,
        options.projectDetails,
      );
      return changed ? this.createGeometryFromWasm() : undefined;
    } finally {
      buffers.positions.delete();
      buffers.indices.delete();
    }
  }

  previewPixRemeshTriangles(
    positions: number[],
    indices: number[],
    options: { resolution: number; adaptiveDensity: number; preserveSharpFeatures: boolean; smoothIterations: number; projectDetails: boolean },
  ): number | undefined {
    if (!this.wasm || !this.wasmModule) return undefined;
    const buffers = this.createWasmBuffers(positions, indices);
    try {
      return this.wasm.previewPixRemeshTriangles(
        buffers.positions,
        buffers.indices,
        options.resolution,
        options.adaptiveDensity,
        options.preserveSharpFeatures,
        options.smoothIterations,
        options.projectDetails,
      );
    } finally {
      buffers.positions.delete();
      buffers.indices.delete();
    }
  }

  applyDraw(center: THREE.Vector3, settings: BrushSettings): boolean {
    const wasm = this.wasmForCurrentGeometry();
    if (wasm) {
      const changed = wasm.applyDraw(center.x, center.y, center.z, settings.radius, settings.strength, settings.invert);
      if (changed) this.syncPositionsAndNormals();
      return changed;
    }
    const geometry = this.getGeometry();
    const candidates = queryBrushCandidates(geometry, center, settings.radius);
    geometry.userData.lastBrushCandidates = candidates;
    return this.drawFallback.apply({
      geometry,
      center,
      settings,
      neighbors: buildVertexNeighbors(geometry),
      candidates,
    });
  }

  applySmooth(center: THREE.Vector3, settings: BrushSettings): boolean {
    const wasm = this.wasmForCurrentGeometry();
    if (wasm) {
      const changed = wasm.applySmooth(center.x, center.y, center.z, settings.radius, settings.strength);
      if (changed) this.syncPositionsAndNormals();
      return changed;
    }
    const geometry = this.getGeometry();
    const candidates = queryBrushCandidates(geometry, center, settings.radius);
    geometry.userData.lastBrushCandidates = candidates;
    return this.smoothFallback.apply({
      geometry,
      center,
      settings,
      neighbors: buildVertexNeighbors(geometry),
      candidates,
    });
  }

  applyClay(center: THREE.Vector3, planeNormal: THREE.Vector3, settings: BrushSettings): boolean {
    const wasm = this.wasmForCurrentGeometry();
    if (wasm) {
      const changed = wasm.applyClay(
        center.x, center.y, center.z,
        planeNormal.x, planeNormal.y, planeNormal.z,
        settings.radius, settings.strength, settings.invert,
      );
      if (changed) this.syncPositionsAndNormals();
      return changed;
    }
    const geometry = this.getGeometry();
    const candidates = queryBrushCandidates(geometry, center, settings.radius);
    geometry.userData.lastBrushCandidates = candidates;
    return this.clayFallback.applyWithPlane({
      geometry,
      center,
      settings,
      neighbors: buildVertexNeighbors(geometry),
      candidates,
    }, planeNormal);
  }

  beginMove(center: THREE.Vector3, settings: BrushSettings): void {
    const wasm = this.wasmForCurrentGeometry();
    if (wasm) {
      wasm.beginMove(center.x, center.y, center.z, settings.radius);
      return;
    }
    const geometry = this.getGeometry();
    const candidates = queryBrushCandidates(geometry, center, settings.radius);
    this.moveFallback.begin({
      geometry,
      center,
      settings,
      neighbors: buildVertexNeighbors(geometry),
      candidates,
    });
  }

  applyMove(delta: THREE.Vector3, settings: BrushSettings): boolean {
    const wasm = this.wasmForCurrentGeometry();
    if (wasm) {
      const changed = wasm.applyMove(delta.x, delta.y, delta.z, settings.invert);
      if (changed) this.syncPositionsAndNormals();
      return changed;
    }
    return this.moveFallback.applyMove(this.getGeometry(), delta, settings);
  }

  private async load(): Promise<void> {
    try {
      const moduleUrl = `/wasm/sculpt-engine.js?v=${Date.now()}`;
      const nativeImport = new Function('moduleUrl', 'return import(moduleUrl)') as (url: string) => Promise<{ default: () => Promise<WasmModule> }>;
      const createModule = (await nativeImport(moduleUrl)).default;
      const module = await createModule();
      this.wasmModule = module;
      this.wasm = new module.SculptEngine();
      this.loadError = undefined;
      const options = this.getGeometry().userData.options as { radius: number; subdivisions: number } | undefined;
      if (options) this.wasm.createQuadSphere(options.radius, options.subdivisions);
      (globalThis as { __pixWasmReady?: boolean; __pixWasmError?: string }).__pixWasmReady = true;
      (globalThis as { __pixWasmReady?: boolean; __pixWasmError?: string }).__pixWasmError = undefined;
      console.info('Sculpt core: C++ WebAssembly');
      this.onReady?.();
    } catch (error) {
      this.wasm = undefined;
      this.wasmModule = undefined;
      this.loadError = error instanceof Error ? `${error.name}: ${error.message}` : String(error);
      (globalThis as { __pixWasmReady?: boolean; __pixWasmError?: string }).__pixWasmReady = false;
      (globalThis as { __pixWasmReady?: boolean; __pixWasmError?: string }).__pixWasmError = this.loadError;
      console.warn('Sculpt core: TypeScript fallback. Run npm run build:wasm after activating Emscripten.', error);
    }
  }

  private createWasmBuffers(positions: number[], indices: number[]): { positions: MutableEmbindVector<number>; indices: MutableEmbindVector<number> } {
    if (!this.wasmModule) throw new Error('WASM module is not loaded.');
    const wasmPositions = new this.wasmModule.FloatVector();
    const wasmIndices = new this.wasmModule.UIntVector();
    for (const value of positions) wasmPositions.push_back(value);
    for (const value of indices) wasmIndices.push_back(value);
    return { positions: wasmPositions, indices: wasmIndices };
  }

  private syncPositionsAndNormals(): void {
    if (!this.wasm) return;
    const geometry = this.getGeometry();
    this.copyVector(this.wasm.positions(), geometry.getAttribute('position') as THREE.BufferAttribute);
    this.copyVector(this.wasm.normals(), geometry.getAttribute('normal') as THREE.BufferAttribute);
  }

  private createGeometryFromWasm(): THREE.BufferGeometry {
    if (!this.wasm) throw new Error('WASM sculpt engine is not loaded.');
    const geometry = new THREE.BufferGeometry();
    geometry.setAttribute('position', new THREE.Float32BufferAttribute(this.vectorToArray(this.wasm.positions()), 3));
    geometry.setAttribute('normal', new THREE.Float32BufferAttribute(this.vectorToArray(this.wasm.normals()), 3));
    geometry.setIndex(this.vectorToArray(this.wasm.indices()));
    geometry.userData.primitive = 'pix-remesh';
    const quads = this.vectorToArray(this.wasm.quads());
    const displayEdges = this.vectorToArray(this.wasm.displayEdges());
    geometry.userData.wireframeLevels = [new Uint32Array(), new Uint32Array(), new Uint32Array(displayEdges)];
    if (quads.length > 0) {
      const quadFaces = this.flatQuadsToFaces(quads);
      geometry.userData.quadFaces = quadFaces;
    }
    geometry.computeBoundingBox();
    geometry.computeBoundingSphere();
    return geometry;
  }

  private wasmForCurrentGeometry(): WasmEngine | undefined {
    return undefined;
  }

  private canSyncCurrentGeometry(): boolean {
    if (!this.wasm) return false;
    const positions = this.wasm.positions();
    const wasmSize = positions.size();
    positions.delete();
    return wasmSize === (this.getGeometry().getAttribute('position').array as Float32Array).length;
  }

  private copyVector(source: EmbindVector<number>, target: THREE.BufferAttribute): void {
    const array = target.array as Float32Array;
    if (source.size() !== array.length) {
      source.delete();
      throw new Error('WASM and Three.js buffers have different sizes.');
    }
    for (let index = 0; index < array.length; index += 1) array[index] = source.get(index);
    source.delete();
    target.needsUpdate = true;
  }

  private vectorToArray(source: EmbindVector<number>): number[] {
    const result: number[] = [];
    result.length = source.size();
    for (let index = 0; index < result.length; index += 1) result[index] = source.get(index);
    source.delete();
    return result;
  }

  private flatQuadsToFaces(values: number[]): number[][] {
    const faces: number[][] = [];
    for (let index = 0; index + 3 < values.length; index += 4) {
      faces.push([values[index], values[index + 1], values[index + 2], values[index + 3]]);
    }
    return faces;
  }
}
